### PR TITLE
Router component

### PR DESCRIPTION
**Pull request by roleyfoley**
## Description
Adds support for a new router component and support for the gateway to use the router as a destination 

Also includes a fix to the Subnet calculation process, the current process only supports tier counts with perfect square values, this fix uses the closet perfect square for tier counts which don't map to a perfect square. 

## Motivation and Context
The router component is used to provision network traffic between different network components. This allows for private networking and propagated routes to be used when you want to build a mesh based network 

- A router is deployed by itself and by default is not connected to anything 
- A network is deployed by itself and by default is not connected to anything 
- A gateway is deployed to a network based on the tier the gateway belongs to, and its destinations determine which network paths are available to that network. If a gateway is configured with the router engine it needs to link to the router component. This establishes the full path between the components 
- Other gateways can then link to the same route to enable routing between networks 

## How Has This Been Tested?
Tested on local deployment

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves the structure or operation of the implementation)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Followup Actions
- [x] None

## Checklist:
- [ ] My change requires a change to the [documentation](https://github.com/hamlet-io/docs).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [X] None of the above.

### CODE REVIEW

File: engine/base.ftl [hunk 1]
```diff
@@ -71,11 +71,20 @@
 [/#function]
 
 [#function intSquareRoot value ]
+    [#local lastPerfect = 1 ]
     [#list 1..100 as square]
```
Comment (RossMurr4y) [inline engine/base.ftl:75]: It looks like break directives in lists are [deprecated](https://freemarker.apache.org/docs/ref_directive_list.html#ref.directive.list.break). Should we try and avoid that? It recommends using `?take_while()` instead. Something like:

```
[#list (1..100)?take_while(s -> ((s * s) < value)) as square]
```

File: engine/base.ftl [hunk 2]
```diff
@@ -71,11 +71,20 @@
...
-        [#if square * square == value ]
+        [#local multiple = square * square ]
+        [#if multiple == value ]
             [#return square ]
         [/#if]
+        [#if multiple?floor == multiple ]
```
Comment (ml019) [inline engine/base.ftl:80]: value is an integer so when will multiple?floor not equal multiple?

Am I correct in that the function will now return the square root of the smallest perfect square greater than or equal to the provided value?

File: engine/base.ftl (hunk 1 above)
Comment (roleyfoley) [reply engine/base.ftl:77]: Couldn't get this to work, still seems to be missing the value when it goes to the next one up.

Comment (roleyfoley) [reply engine/base.ftl:77]: Couldn't get this to work, still seems to be missing the value when it goes to the next one up.

File: engine/base.ftl (hunk 2 above)
Comment (roleyfoley) [reply engine/base.ftl]: Good point, removed the floor check 

Yes the idea is to find the closest perfect square which can then be used for the smallest subnet mask that we need. I've renamed the function and added a comment

File: engine/base.ftl (hunk 1 above)
Comment (RossMurr4y) [reply engine/base.ftl:77]: Not really an issue then, just something I picked out as I hadn't seen a `break` used here before.

**Pull request merged**